### PR TITLE
fix(nuxt): ensure `error` in `useAsyncData` has correct type

### PR DESCRIPTION
**Pull request by DamianGlowala**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [x] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Updates error handling composables and defines asyncData's error as `NuxtError` with generic Nuxt error's `data` property.
It also avoids `as any` casting in `createError()` to define `__nuxt_error` by using the `Object.defineProperty`.

(Will double check the docs for relevant updates once it's confirmed this update is good to go.)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt/src/app/composables/asyncData.ts [hunk 1]
```diff
@@ -74,53 +75,53 @@ export type AsyncData<Data, Error> = _AsyncData<Data, Error> & Promise<_AsyncDat
 
 export function useAsyncData<
   ResT,
-  DataE = Error,
+  NuxtErrorDataT = any,
```
Comment (danielroe) [inline packages/nuxt/src/app/composables/asyncData.ts:78]: what about defaulting to `unknown`?

Comment (DamianGlowala) [reply packages/nuxt/src/app/composables/asyncData.ts:78]: Sure. Should we consider defaulting to `unknown` directly in h3 (https://github.com/unjs/h3/blob/main/src/error.ts) as well?

Comment (danielroe) [reply packages/nuxt/src/app/composables/asyncData.ts]: I would also recommend that but I know @pi0 may have a different timetable or opinion on this one.

**Review by danielroe — COMMENTED** [review]
This seems like a good change but is also a breaking one for users specifying an error type.

See for example: https://github.com/search?q=/use(Lazy)?(AsyncData%7CFetch)+?%3C%5B%5E%3E,%5D,%5B%5E%3E%5D/&type=code

We might use inference to return previous return value if `NuxtErrorDataT` extends `Error` or `NuxtError`. wdyt?

**Comment by DamianGlowala** [comment]
Yes, I agree, good find. Majority of the linked code snippets are React-related, but there were certainly a few Nuxt-related ones in these search results too.

Would the following be fine instead of `NuxtError<NuxtErrorDataT> | null`?
```
(NuxtErrorDataT extends Error | NuxtError ? NuxtErrorDataT : NuxtError<NuxtErrorDataT>) | null
```

**Pull request merged**